### PR TITLE
DAOS-1574 control: return numa node for nvme on scan

### PR DESCRIPTION
**Pull request by tanabarr**
Add NUMA node/socket ID for NVMe SSDs to details returned from
storage scan.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by tanabarr** [comment]
relevant output
`
- model: 'INTEL SSDPEDMD016T4 '
  serial: 'CVFT5226001C1P6DGN  '
  pciaddr: 0000:87:00.0
  fwrev: 8DV10171
  numanode: 0
  namespaces:
  - id: 1
    capacity: 1600
`

**Comment by tanabarr** [comment]
updated output, consistent between client and server commands:
```
[root@wolf-72 daos_m]# install/bin/daos_server storage scan                                                    Scanning locally-attached storage...
Starting SPDK v18.07-pre / DPDK 18.02.0 initialization...
[ DPDK EAL parameters: spdk -c 0x1 --file-prefix=spdk_pid69300 ]
EAL: Detected 96 lcore(s)
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Multi-process socket /var/run/.spdk_pid69300_unix
EAL: Probing VFIO support...
EAL: PCI device 0000:81:00.0 on NUMA socket 1
EAL:   probe driver: 8086:2701 spdk_nvme
EAL: PCI device 0000:87:00.0 on NUMA socket 1
EAL:   probe driver: 8086:953 spdk_nvme
EAL: PCI device 0000:da:00.0 on NUMA socket 1
EAL:   probe driver: 8086:2701 spdk_nvme

NVMe SSD controller and constituent namespaces:
        PCI Address:0000:da:00.0 Serial:PHKS7505005Y750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
        PCI Address:0000:81:00.0 Serial:PHKS7505007J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
        PCI Address:0000:87:00.0 Serial:CVFT5392000G1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket: 0
                Namespace: id:1 capacity:1600

SCM modules:
        Physical ID: 40 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 1 pos 1)
        Physical ID: 44 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 2 pos 1)
        Physical ID: 36 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 0 pos 1)
        Physical ID: 52 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 1 pos 0)
        Physical ID: 55 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 2 pos 0)
        Physical ID: 50 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 0 pos 1)
        Physical ID: 62 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 0 pos 1)
        Physical ID: 66 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 1 pos 1)
        Physical ID: 70 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 2 pos 1)
        Physical ID: 76 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 0 pos 1)
        Physical ID: 78 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 1 pos 0)
        Physical ID: 81 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 2 pos 0)
```
and
```
[tanabarr@wolf-72 daos_m]$ install/bin/daos_shell -l wolf-72:10001 -i storage scan
Active connections: [wolf-72:10001]
NVMe SSD controller and constituent namespaces:
wolf-72:10001:
        PCI Address:0000:da:00.0 Serial:PHKS7505005Y750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
        PCI Address:0000:81:00.0 Serial:PHKS7505007J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
        PCI Address:0000:87:00.0 Serial:CVFT5392000G1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket: 0
                Namespace: id:1 capacity:1600

SCM modules:
wolf-72:10001:
        Physical ID: 40 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 1 pos 1)
        Physical ID: 44 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 2 pos 1)
        Physical ID: 36 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 0 pos 1)
        Physical ID: 52 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 1 pos 0)
        Physical ID: 55 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 2 pos 0)
        Physical ID: 50 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 0 pos 1)
        Physical ID: 62 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 0 pos 1)
        Physical ID: 66 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 1 pos 1)
        Physical ID: 70 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 2 pos 1)
        Physical ID: 76 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 0 pos 1)
        Physical ID: 78 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 1 pos 0)
        Physical ID: 81 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 2 pos 0)
```

**Comment by mjmac** [comment]
I'll take a look in depth too, but just looking at the example output I wonder what you think about the option of grouping devices by numa node in sort of a tree view? Maybe in a follow on PR to avoid scope creep on this one...

e.g.
```
wolf-72:10001:
    Socket 0:
        NVMe:
            PCI Address:0000:da:00.0 Serial:PHKS7505005Y750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
            PCI Address:0000:81:00.0 Serial:PHKS7505007J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
                Namespace: id:1 capacity:750
            PCI Address:0000:87:00.0 Serial:CVFT5392000G1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket: 0
                Namespace: id:1 capacity:1600
        SCM:
            Physical ID: 40 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 1 pos 1)
            Physical ID: 44 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 2 pos 1)
            Physical ID: 36 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 0 pos 1)
            Physical ID: 52 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 1 pos 0)
            Physical ID: 55 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 2 pos 0)
            Physical ID: 50 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 0 pos 1)
    Socket 1:
        NVMe:
        SCM:
            Physical ID: 62 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 0 pos 1)
            Physical ID: 66 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 1 pos 1)
            Physical ID: 70 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 2 pos 1)
            Physical ID: 76 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 0 pos 1)
            Physical ID: 78 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 1 pos 0)
            Physical ID: 81 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 2 pos 0) 
```

File: src/control/common/storage_types.go [hunk 1]
```diff
@@ -38,13 +38,13 @@ func (nc NvmeControllers) String() string {
 	var buf bytes.Buffer
 
 	for _, ctrlr := range nc {
-		fmt.Fprintf(
-			&buf, "\tPCI Address:%s Serial:%s\n\tModel:%s Fwrev:%s\n",
-			ctrlr.Pciaddr, ctrlr.Serial, ctrlr.Model, ctrlr.Fwrev)
+		fmt.Fprintf(&buf,
+			"\tPCI Address:%s Serial:%s Model:%s Fwrev:%s Socket: %d\n",
```
Comment (mjmac) [inline src/control/common/storage_types.go]: One concern I have with this formatting choice is that the line break makes it harder to run the output through grep. I'd recommend removing any mid-entry line breaks, even if it results in longer output lines.

File: src/control/common/storage_types.go [hunk 2]
```diff
@@ -38,13 +38,13 @@ func (nc NvmeControllers) String() string {
 	var buf bytes.Buffer
 
 	for _, ctrlr := range nc {
-		fmt.Fprintf(
-			&buf, "\tPCI Address:%s Serial:%s\n\tModel:%s Fwrev:%s\n",
```
Comment (mjmac) [inline src/control/common/storage_types.go]: Looking at this a bit more, I also notice an inconsistency between the SCM formatting and NVMe formatting. Here, the fields are formatted with `$name:%s`, but in SCM it's `$name: %s`. It's a minor thing visually, but irritating to deal with when trying to parse output with other tools.

Comment (tanabarr) [reply src/control/common/storage_types.go]: I noticed that as well just after pushing, will fix as I agree it's irritating

File: src/control/common/storage_types.go (hunk 1 above)
Comment (tanabarr) [reply src/control/common/storage_types.go]: agreed, presumably the json option could be used for that

**Comment by tanabarr** [comment]
> I'll take a look in depth too, but just looking at the example output I wonder what you think about the option of grouping devices by numa node in sort of a tree view? Maybe in a follow on PR to avoid scope creep on this one...
> 
> e.g.
> 
> ```
> wolf-72:10001:
>     Socket 0:
>         NVMe:
>             PCI Address:0000:da:00.0 Serial:PHKS7505005Y750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
>                 Namespace: id:1 capacity:750
>             PCI Address:0000:81:00.0 Serial:PHKS7505007J750BGN   Model:INTEL SSDPED1K750GA  Fwrev:E2010325 Socket: 0
>                 Namespace: id:1 capacity:750
>             PCI Address:0000:87:00.0 Serial:CVFT5392000G1P6DGN   Model:INTEL SSDPEDMD016T4  Fwrev:8DV10171 Socket: 0
>                 Namespace: id:1 capacity:1600
>         SCM:
>             Physical ID: 40 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 1 pos 1)
>             Physical ID: 44 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 2 pos 1)
>             Physical ID: 36 Capacity: 539661172736 Location: (socket 0 memctrlr 0 chan 0 pos 1)
>             Physical ID: 52 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 1 pos 0)
>             Physical ID: 55 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 2 pos 0)
>             Physical ID: 50 Capacity: 539661172736 Location: (socket 0 memctrlr 1 chan 0 pos 1)
>     Socket 1:
>         NVMe:
>         SCM:
>             Physical ID: 62 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 0 pos 1)
>             Physical ID: 66 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 1 pos 1)
>             Physical ID: 70 Capacity: 539661172736 Location: (socket 1 memctrlr 0 chan 2 pos 1)
>             Physical ID: 76 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 0 pos 1)
>             Physical ID: 78 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 1 pos 0)
>             Physical ID: 81 Capacity: 539661172736 Location: (socket 1 memctrlr 1 chan 2 pos 0) 
> ```

I would be fine with that and agree that it probably makes sense grouping by socket

**Pull request merged**